### ISSUE TITLE
opened modules don't work in datatype reveals

### ISSUE THREAD

**Issue by seanmcl** [dafny-lang/dafny#2028]
```
module M {
  export
    provides
      Opened
    reveals
      D

  module Opened {
    type T = int
  }

  import opened Opened

  datatype D = D(x: T)
}
```

```
 Error: Raised while checking export set M: Undeclared top-level type or type parameter: T (did you forget to qualify a name or declare a module import 'opened'
```

Works fine moving `D` to `provides`.

**Comment by atomb** [comment]
I believe that this is operating as designed (according to [this paper](http://leino.science/papers/krml261.pdf), especially section 3.3), but that behavior is clearly inconvenient in this case.

Inspired by that paper section, I tried the following, which does work:

````
module M {
  export
    provides
      Opened, U
    reveals
      D

  module Opened {
    type T = int
  }

  import opened Opened

  type U = T

  datatype D = D(x: U)
}
````

Would structuring things this way be a viable solution in the case you're working with?

**Comment by seanmcl** [comment]
Thanks. Renaming all the types is not an option. What doesn't make sense to me here is that if I remove the export clause everything works just fine. I maintain, for that reason, that this is a bug.

**Comment by atomb** [comment]
Besides the acceptance/rejection of any specific example, the error message Dafny produces in this case doesn't clearly point to a solution and could perhaps be improved.

**Comment by atomb** [comment]
Ah, yes, you're right! The behavior with the default (implicit) export set doesn't seem to be possible with an explicit export set.

**Comment by atomb** [comment]
I think @RustanLeino may be the most qualified to understand what's going on here.

**Comment by atomb** [comment]
Interesting. The following works, but not if you remove `T` from the `provides` set of `M`.

````
module M {
  export
    provides
      Opened, T
    reveals
      D

  module Opened {
    type T = int
  }

  import opened Opened

  type T = Opened.T

  datatype D = D(x: T)
}

module N {
  import M
  type V = M.D
  type S = M.Opened.T
  type T = M.T
}
````

One thing to note about this example, though: it allows the use of `T` without changing its unqualified name, though it still requires the introduction of a (somewhat meaningless?) alias.

**Comment by RustanLeino** [comment]
Evidently, exports of imports and exports of nested modules have not received the attention they need (in the language design and implementation). To explain what (in my opinion) _should_ happen, let me first talk to an example that uses an `import` and then talk to an example that uses a nested `module`.

Consider the following program:

``` dafny
module Library {
  export JustA
    reveals A
  export BothAB
    provides A, B

  type A = int
  type B = real
}

module M {
  export
    provides LA, F, G

  import LA = Library`JustA
  import LAB = Library`BothAB

  function F(x: LA.A, y: LAB.A): bool // (0)
    requires y == 7 // (1)
    requires x == y // (2)

  function G(w: LAB.A, z: LAB.B): bool // (3)
}
```

0. The signature of function `F` is exported by module `M`. The signature literally mentions both `LA` and `LAB`, whereas the export set mentions only `LA`. Dafny currently allows this, but I think this ought to be _disallowed_. (An alternative design would be to allow it, since, even though the name `LAB` is not exported, `LAB.A` and `LA.A` refer to the same type and `LA` is exported. If someone likes this alternative design and would like to argue for it, please speak up.)
1. Since `M`'s export set provides `F`, the export set must include enough things so that `x == 7` makes sense. This is indeed the case in the example, since the export set includes ``Library`JustA``, which reveals that `A` is the type `int`. If `reveals A` in module `Library` is changed to just `provides A`, then line (1) should give an error (which is currently what happens). Curiously, if `Library`'s export sets are changed to
    ``` dafny
    export JustA
      provides A
    export BothAB
      reveals A, B
    ```
    Dafny currently reports an error that line (1) makes `M` export set inconsistent (good!). So, apparently, in the current implementation, Dafny allows `LAB.A` to be mentioned even if the name `LAB` is not exported, but it does not recognize that `LAB.A` and `LA.A` are the same type.
2. Since `M`'s export set provides `F`, the export set must include enough things so that `x == y` makes sense. This is currently allowed, since `LA.A` and `LAB.A` refer to the same type. However, I think this ought to be _disallowed_ (because I think line (0) should not be allowed to literally use `LAB`).
3. Line (3) mentions `LAB.B`. This should be disallowed, since `LAB` is not exported. I argue the situation in (3) is the same as the situation in (0), but in (3) it is clear that `LAB.B` is not something that's exported by `M` (unlike in (0), where one _could_ argue that the mention `LAB.A` should be allowed, since `LAB.A` and `LA.A` refer to the same type).

So, as it stands, Dafny currently ignores which `Library` export set is exported by `M`--apparently, everything `M` imports about `Library` gets exported when `M` exports any name (like `LA`) that refers to one of the export sets of `Library`. That seems wrong to me.

What I've said here about exporting the name `LA`, which is `M`'s local name for a module, ought to be the same even if the module to which `LA` refers is declared inside `M` itself. Currently, if you move module `Library` to sit inside module `M`, which means that the `import ... = Library...` declarations are referring to a nested module, then the four bullets above still apply. Indeed, Dafny seems to treat these imports the same, regardless of where `Library` is declared.

If `Library` is declared inside `M`, then it ought to be fine for `M` to export `Library`, just as if it had declared `import L = Library` and exported `L`. This is broken today, because the following program results in errors:

``` dafny
module M {
  export
    provides Library, H

  function H(w: Library.A): bool // BOGUS: an error is reported here, saying export set is not consistent

  module Library {
    export
      reveals A

    type A = int
  }
}
```

The same error is bogusly reported if the `export` declaration is changed to:

``` dafny
  export
    reveals *
```

Curiously, by removing the `export` declaration altogether, which semantically defaults to `export reveals *`, Dafny reports no errors.

In summary, when an export set (like the one in my first example above) provides a name (like `LA`) that is the local name for an imported module or a nested module, then the export set should only be exporting those things in the module that the name `LA` refers to. And no export set at all should always be the same as `export reveals *`.
